content/blog/2023/2023-06-25-archunit.md

```diff
@@ -0,0 +1,598 @@
+time we might forget. Also, other developers who join the project might unintentionally deviate from the original choice. 
+
+With ArchUnit, we can add a test that will fail if the static factory method ```now``` is used without the parameter: 
+
+```java
+@Test
+public void alwayInjectClockToFaciliateTesting() { 
+  // ...
+}
+```
+
+Such a test reminds the developer to use the parameter and remain consistent within the codebase. It also explains the reason
+why to use (or not to use) a specific method.
+
+This is a good example of how we can use ArchUnit to document the intended architecture in the form of unit tests. Of course,
+we can change things when we see the need for it. There might be good reasons to deviate from a certain pattern. However,
+using tests as a documentation will remind us to think about why we want to deviate.
+
+## ArchUnit Examples 
+<a name="examples"></a>
+Most examples of how to use ArchUnit describe checks on dependencies between classes and packages.
+That's however not the only use case. Let's have a look at three other types of checks that we can create.
+
+### @deprecated
+```java
+@Test
+public void do_not_call_deprecated_methods_from_the_project() {
+  JavaClasses importedClasses = new ClassFileImporter().importPackages("io.reflectoring.archunit");
+  ArchRule rule = noClasses().should().dependOnClassesThat().areAnnotatedWith(Deprecated.class);
+  rule.check(importedClasses);
+}
+
+public void reference_deprecated_class() {
+  Dep dep = new Dep();
+}
+
+@Deprecated
+public class Dep {
+}
+```
+With this test, we can check if we still use any deprecated methods. This check can be very useful in refactoring
+projects where we want to upgrade the version of libraries.
+
+### BigDecimal 
+Another nice use case is to **prevent the use of a specific constructor of a class**. Why would we won't to do this?
+IDEs usually show us a warning (including an explanation) and code quality tools such as SonarQube can be configured 
+to detect these cases as well.
+
+With ArchUnit, we can easily achieve this in a unit test, which makes our intention to exclude a certain constructor
+clear. It reminds us that we really do not want to use a specific method or constructor call and we'll get a
+failed test instead of only a warning. Another benefit is, that with ArchUnit, we can introduce custom checks.
+
+As an example, let's see how we can prevent calling one of the constructors of the `BigDecimal` class:
+
+```java
+@Test
+public void do_not_call_constructor() {
+  JavaClasses importedClasses = new ClassFileImporter().importPackages("io.reflectoring.archunit");
+  ArchRule rule = noClasses().should().callConstructor(BigDecimal.class, double.class);
+  rule.check(importedClasses);
+}
+```
+    
+This test will fail if we call the BigDecimal constructor that accepts a double value as a parameter:
+```java
+public void this_method_calls_the_wrong_BigDecimal_constructor() {
```

Comment:
Is there a reason why you use the snake_case and not the camelCase

Reply:
No, I have updated all method names.